### DISCUSSION TITLE
code navigation

### DISCUSSION THREAD

**Discussion by truher** [community/community#33625] [Codespaces]
hi, i'm using github.dev codespaces for editing and compiling java and it works fine, but none of the vscode navigation features seem to exist, e.g. "go to definition" or "find all references," stuff like that.   are these features supposed to exist?  is there some extension that can provide them?  thanks!

**Comment by bjbennet** [comment]
Hi @truher , generally speaking these language specific services will not be available. Specifically for the Java language, it needs the full language server to provide intellisense and "go to definition" support. That service cannot run in the client-side browser.

**Reply by truher** [reply]
thanks for your reply! i was hoping that the editor could make use of the full language server in the attached codespaces docker VM ... the thing it uses for compiling. is that not the case? is this the sort of thing that a web extension might be able to do?

**Reply by cdwort** [reply]
Hi @truher the VS Code editor in a full codespace absolutely can use the compute provisioned! However, our github.dev editing experience runs fully in the browser to help enable quick edits and file navigation. It does not have any compute -- there is no VM nor docker container running.

**Reply by truher** [reply]
then maybe i just have it misconfigured somehow?  i'm running the codespace thing, but i'm not seeing any effect of that in the editor. is there an example of what you're talking about that i could reference?

**Reply by cdwort** [reply]
Hi @truher, yes the distinction between Codespaces and GitHub.dev isn't the most clear. 

In a GitHub.dev editor, you will not see indications of an attached VM. You can see there's no terminal option here: 
<img width="255" alt="Screen Shot 2022-09-22 at 6 44 12 AM" src="https://user-images.githubusercontent.com/1014482/191739114-9cd63df1-bb81-4c09-bd11-33a1273586cb.png">
You can also tell you're in the web-based editor by the url matching the repo url except for three characters, e.g.:
```
github.com/truher/main2023
github.dev/truher/main2023
```

In a Codespace, you will see that there is a VM:
<img width="224" alt="Screen Shot 2022-09-22 at 6 45 19 AM" src="https://user-images.githubusercontent.com/1014482/191739563-545ad6aa-66d4-42a2-8c42-d92dc49c3969.png">
And the url will have the codespace name it in, not the repo name, e.g. `https://cdwort-truher-main2023-5g4vv46gg43p75r.github.dev/`. 

I know that having codespaces on the `github.dev` domain now makes things even more confusing.

You can read more here: https://docs.github.com/en/codespaces/the-githubdev-web-based-editor#codespaces-and-the-web-based-editor.

**Reply by truher** [reply]
hm, ok, thanks for your speedy reply.  it definitely didn't work before, and i was definitely using codespaces (it took awhile to build the docker container, so i remember), but it's definitely working now.  anyway, thanks, looks great!

**Comment by snebjorn** [comment]
I have a similar issue. JAVA code navigation doesn't work and I'm using GitHub Codespaces in my local VSCode.
My `devcontainer.json` have the following extensions enabled
```jsonc
  "customizations": {
    // Configure properties specific to VS Code.
    "vscode": {
      // Set *default* container specific settings.json values on container create.
      "settings": {},
      // Add the IDs of extensions you want installed when the container is created.
      "extensions": [
        "vscjava.vscode-java-pack",
        "attilabuti.mustache-syntax-vscode",
        "formulahendry.code-runner",
        "visualstudioexptteam.vscodeintellicode",
        "42crunch.vscode-openapi",
        "mermade.openapi-lint"
      ]
    }
  },
```
![image](https://user-images.githubusercontent.com/1266245/193657153-404fd20d-1032-4943-ae1e-dad33427714c.png)
![image](https://user-images.githubusercontent.com/1266245/193672215-775e0d15-481e-4e9f-b71a-f591863d270f.png)

Java is installed and setup in my container
```bash
$ java -version
openjdk version "11.0.16.1" 2022-08-12 LTS
OpenJDK Runtime Environment Microsoft-40648 (build 11.0.16.1+1-LTS)
OpenJDK 64-Bit Server VM Microsoft-40648 (build 11.0.16.1+1-LTS, mixed mode)
```

But code navigation or language support doesn't seem to work :(
